File: dynamic-programming/climbing-n-stairs.py
# 1, 2, 3, ... 

# dynamic programming approach (top down)
# O(n) time | O(n) space
def climbingNStiars(n, memo = {1: 1, 2: 2}):
    if n in memo:
        return memo[n]
    else:
        memo[n] = climbingNStiars(n-1, memo) + climbingNStiars(n-2, memo)
        return memo[n]

print(climbingNStiars(6))


# iterative solution 1 (bottom up)
# O(n) time | O(n) space
def climbingNStiars_iterative(n):
    table = [None]*(n + 1)
    table[1] = 1
    table[2] = 2
    for i in range(3, n + 1):
        table[i] = table[i - 1] + table[i - 2]
    return table[n]

print(climbingNStiars_iterative(6))

# iterative solution 2
# O(n) time | O(1) space
def climbingNStiars_iterative_2(n):
    lastTwo = [1, 2]
    counter = 3
    while counter <= n:
        nextN = lastTwo[0] + lastTwo[1]
        lastTwo[0] = lastTwo[1]
        lastTwo[1] = nextN
        counter += 1
    if n > 1:
        return lastTwo[1]
    else:
        return lastTwo[0]

print(climbingNStiars_iterative_2(6))

File: js/dynamic programming/climbingNStairs.js
/**
 * time-complexity:
 *   - O(2^n)
 *
 * space-complexity:
 *   - O(n)
 *
 * @param {*} n
 */
function climbNStairs(n) {
  if (n === 1 || n === 2) return n;

  return climbNStairs(n - 1) + climbNStairs(n - 2);
}

/**
 * TOP DOWN DP APPROACH
 *
 * time-complexity:
 *   - O(n)
 *
 * space-complexity:
 *   - O(n)
 *
 * @param {*} n
 */
function climbNStairsDP(n, cache = { 1: 1, 2: 2 }) {
  if (n < 1) return -1;

  if (n in cache) {
    return cache[n];
  }

  cache[n] = climbNStairsDP(n - 1, cache) + climbNStairsDP(n - 2, cache);
  return cache[n];
}

/**
 * BOTTOM UP DP APPROACH
 *
 * time-complexity:
 *   - O(n)
 *
 * space-complexity:
 *   - O(1)
 *
 * @param {*} n
 */
function climbNStairsBottomUp(n) {
  const lastTwo = [1, 2];
  let counter = 3;

  while (counter <= n) {
    const temp = lastTwo.reduce((a, b) => a + b);
    lastTwo[0] = lastTwo[1];
    lastTwo[1] = temp;
    counter++;
  }

  if (n > 1) {
    return lastTwo[1];
  }

  return lastTwo[0];
}

console.log(climbNStairsBottomUp(4));
